cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/TaskWorkerService.java

```diff
@@ -0,0 +1,118 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.worker;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.conf.SConfiguration;
+import io.cdap.cdap.common.discovery.ResolvingDiscoverable;
+import io.cdap.cdap.common.discovery.URIScheme;
+import io.cdap.cdap.common.http.CommonNettyHttpServiceBuilder;
+import io.cdap.cdap.common.internal.worker.RunnableTask;
+import io.cdap.cdap.common.security.HttpsEnabler;
+import io.cdap.http.NettyHttpService;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.discovery.DiscoveryService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Launches an HTTP server for receiving and handling {@link RunnableTask}
+ */
+public class TaskWorkerService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TaskWorkerService.class);
+
+  private final CConfiguration cConf;
+  private final SConfiguration sConf;
+  private final DiscoveryService discoveryService;
+  private final NettyHttpService httpService;
+  private Cancellable cancelDiscovery;
+  private InetSocketAddress bindAddress;
+
+  @Inject
+  TaskWorkerService(CConfiguration cConf,
+                    SConfiguration sConf,
+                    DiscoveryService discoveryService) {
+    this.cConf = cConf;
+    this.sConf = sConf;
+    this.discoveryService = discoveryService;
+
+    NettyHttpService.Builder builder = new CommonNettyHttpServiceBuilder(cConf, Constants.Service.TASK_WORKER)
+      .setHost(cConf.get(Constants.TaskWorker.ADDRESS))
+      .setPort(cConf.getInt(Constants.TaskWorker.PORT))
+      .setExecThreadPoolSize(cConf.getInt(Constants.TaskWorker.EXEC_THREADS))
+      .setBossThreadPoolSize(cConf.getInt(Constants.TaskWorker.BOSS_THREADS))
+      .setWorkerThreadPoolSize(cConf.getInt(Constants.TaskWorker.WORKER_THREADS))
+      .setHttpHandlers(new TaskWorkerHttpHandlerInternal(this.cConf, this::stopService));
+
+    if (cConf.getBoolean(Constants.Security.SSL.INTERNAL_ENABLED)) {
+      new HttpsEnabler().configureKeyStore(cConf, sConf).enable(builder);
+    }
+    httpService = builder.build();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.debug("Starting TaskWorker service");
+    launchHttpServer();
+    LOG.debug("Starting TaskWorker service has completed");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.debug("Shutting down TaskWorker service");
+    shutdownHttpServer();
+    LOG.debug("Shutting down TaskWorker service has completed");
+  }
+
+  private void launchHttpServer() throws Exception {
```

Comment:
We can just have this private method inline in the `startUp`.

Reply:
Done!